cmd/remote_val_harness/internal/test_harness.go

```diff
@@ -0,0 +1,391 @@
+	}
+	return nil
+}
+
+// TestSignVote makes sure the remote signer can successfully sign all kinds of
+// votes.
+func (th *TestHarness) TestSignVote() error {
+	th.logger.Info("TEST: Signing of votes")
+	for _, voteType := range voteTypes {
+		th.logger.Info("Testing vote type", "type", voteType)
+		hash := tmhash.Sum([]byte("hash"))
+		vote := &types.Vote{
+			Type:   voteType,
+			Height: 12345,
+			Round:  23456,
+			BlockID: types.BlockID{
+				Hash: hash,
+				PartsHeader: types.PartSetHeader{
+					Hash:  hash,
+					Total: 1000000,
+				},
+			},
+			ValidatorIndex:   0,
+			ValidatorAddress: tmhash.SumTruncated([]byte("addr")),
+			Timestamp:        time.Now(),
+		}
+		voteBytes := vote.SignBytes(th.chainID)
+		// sign the vote
+		if err := th.sc.SignVote(th.chainID, vote); err != nil {
+			th.logger.Error("FAILED: Signing of vote", "err", err)
+			return newTestHarnessError(ErrTestSignVoteFailed, err, fmt.Sprintf("voteType=%d", voteType))
+		}
+		th.logger.Debug("Signed vote", "vote", vote)
+		// validate the contents of the vote
+		if err := vote.ValidateBasic(); err != nil {
+			th.logger.Error("FAILED: Signed vote is invalid", "err", err)
+			return newTestHarnessError(ErrTestSignVoteFailed, err, fmt.Sprintf("voteType=%d", voteType))
+		}
+		// now validate the signature on the proposal
+		if th.sc.GetPubKey().VerifyBytes(voteBytes, vote.Signature) {
+			th.logger.Info("Successfully validated vote signature", "type", voteType)
+		} else {
+			th.logger.Error("FAILED: Vote signature validation failed", "type", voteType)
+			return newTestHarnessError(ErrTestSignVoteFailed, nil, "signature validation failed")
+		}
+	}
+	return nil
+}
+
+// Shutdown will kill the test harness and attempt to close all open sockets
+// gracefully. If the supplied error is nil, it is assumed that the exit code
+// should be 0. If err is not nil, it will exit with an exit code related to the
+// error.
+func (th *TestHarness) Shutdown(err error) {
+	var exitCode int
+
+	if err == nil {
+		exitCode = NoError
+	} else if therr, ok := err.(*TestHarnessError); ok {
+		exitCode = therr.Code
+	} else {
+		exitCode = ErrOther
+	}
+	th.exitCode = exitCode
+
+	// in case sc.Stop() takes too long
+	if th.exitWhenComplete {
+		go func() {
+			time.Sleep(time.Duration(5) * time.Second)
+			th.logger.Error("Forcibly exiting program after timeout")
+			os.Exit(exitCode)
+		}()
+	}
+
+	if th.sc.IsRunning() {
+		if err := th.sc.Stop(); err != nil {
+			th.logger.Error("Failed to cleanly stop listener: %s", err.Error())
+		}
+	}
+
+	if th.exitWhenComplete {
+		os.Exit(exitCode)
+	}
+}
+
+// newTestHarnessSocketVal creates our client instance which we will use for
+// testing.
+func newTestHarnessSocketVal(logger log.Logger, cfg TestHarnessConfig) (*privval.SocketVal, error) {
+	proto, addr := cmn.ProtocolAndAddress(cfg.BindAddr)
+	if proto == "unix" {
+		// make sure the socket doesn't exist - if so, try to delete it
+		if cmn.FileExists(addr) {
```

Comment:
Question: This is the only difference to `newSocketVal` in [client_test.go](https://github.com/tendermint/tendermint/blob/4d7b29cd8f8ef998cf8f7141f3a100b0a4fa718d/privval/client_test.go#L379-L399). 

Did you experience that the socket file was left over from previous test runs here?

Reply:
Yes, I've seen it a few times where, even though we've closed the Unix socket, it still exists in the file system. IIRC, it happened during testing. Checking my `/tmp` folder right now, I still have 58 `.sock` files lying around from testing 😁 

Also, IIRC, I've had it happen at times that if the file already exists, the startup sequence fails because the file exists.